sdks/python/apache_beam/runners/portability/stager.py

```diff
@@ -686,12 +723,49 @@ def _populate_requirements_cache(requirements_file, cache_dir):
           tmp_requirements_filepath,
           '--exists-action',
           'i',
-          # Download from PyPI source distributions.
-          '--no-binary',
-          ':all:'
       ]
-      _LOGGER.info('Executing command: %s', cmd_args)
-      processes.check_output(cmd_args, stderr=processes.STDOUT)
+
+      # Download only sdists to support backward compatibility
+      if not fetch_binary:
+        cmd_args.extend(['--no-binary', ':all:'])
+        _LOGGER.info('Executing command: %s', cmd_args)
+        processes.check_output(cmd_args, stderr=processes.STDOUT)
+      else:
+        try:
+          pip_compile_cmd = [
+              Stager._get_python_executable(),
+              '-m',
+              'piptools',
+              'compile',
+              requirements_file
+          ]
+          processes.check_output(pip_compile_cmd, stderr=processes.STDOUT)
+        except:  # pylint:disable=bare-except
+          # If pip-compile fails, fallback to download of packages with
+          # --no-deps and let worker take care of the transitive deps
+          # of those packagespip-co.
+          _LOGGER.warning(
+              'pip-compile failed. Staging packages from requirements.txt'
```

Comment:
A better warning saying `Downloading packages from requirements_file without transitive/secondary dependencies` or `Staging packages from requirements_file with --no-deps flag`. or may be we skip any warning and just pass the exception and download packages with --no-deps